crates/biome_js_analyze/src/lint/nursery/no_restricted_imports.rs

```diff
@@ -47,6 +58,221 @@ declare_lint_rule! {
+    /// Nevertheless, the rule tries to detect the following common usage patterns where the set of imported names is determined statically:
+    ///
+    /// ```js
+    /// // Dynamic `import()` calls:
+    /// // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import
+    ///
+    /// import('sideeffect-import');
+    /// await import('sideeffect-import');
+    ///
+    /// // ...using await + destructuring-assignment:
+    /// const alias1 = await import('namespace-import');
+    /// const { default: defaultExport } = await import('default-import')
+    /// const { export1, export2: alias2, "string-name": alias3, default: defaultExport /* … */ } = await import("named-import");
+    ///
+    /// // ...using then() with arrow-function + destructuring parameters:
+    /// import('namespace-import').then(alias1 => { /* … */ });
+    /// import('namespace-import').then((alias1) => { /* … */ });
+    /// import('default-import').then(({ default: defaultExport }) => { /* … */ });
+    /// import('named-import').then(({ export1, export2: alias2, "string-name": alias3, default: defaultExport /* … */ }) => { /* … */ });
+    ///
+    /// // ...using then() with function + destructuring parameters:
+    /// import('namespace-import').then(function(alias1) { /* … */ });
+    /// import('default-import').then(function({ default: defaultExport }) { /* … */ });
+    /// import('named-import').then(function({ export1, export2: alias2, "string-name": alias3, default: defaultExport /* … */ }) { /* … */ });
+    ///
+    /// // Standalone `import('...')` calls that appear in some other
+    /// // unrecognized context will be treated as a namespace import,
+    /// // because the return value of `import('...')` is a namespace object:
+    ///
+    /// myFunction(...args, import("namespace-import"), ...args)
+    /// ```
+    ///
+    /// ### Dynamic `require()` calls
+    ///
+    /// NodeJS-style `require()` calls are also supported.
+    /// Due to the way `require()` works, these are always treated as default imports.
+    ///
+    /// ```js
+    /// // Dynamic `require()` call
+    /// const defaultExport = require('default-import');
+    /// ```
+    ///
+    /// ## Options
+    ///
+    /// ```json,options
+    /// {
+    ///     "options": {
+    ///         "paths": {
+    ///             "lodash": "Using lodash is not encouraged",
+    ///             "underscore": "Using underscore is not encouraged",
+    ///             "import-foo": {
+    ///                 "importNames": ["Bar"],
+    ///                 "message": "Please use Bar from /import-bar/baz/ instead."
+    ///             },
+    ///             "import-bar": {
+    ///               "allowImportNames": ["Bar"],
+    ///               "message": "Please use only Bar from import-bar."
+    ///             }
+    ///         }
+    ///     }
+    /// }
+    /// ```
```

Comment:
When explaining options and such, we usually follow this pattern:
1. explain the option
1. explain the example
1. code of the example

This code example is a bit out of context because it isn't followed by any explanation, so it shouldn't be here.